modules/api/VcenterConnection.py

```diff
@@ -0,0 +1,154 @@
+
+    def _get_obj(self, root, vim_type):
+        """
+        Creates a ContainterView and returns it. Can be used to fetch all
+        hosts or vms in one turn.
+
+        :param root: The root where to start. eg vcenter.content.root
+        :param vim_type: vimtype eg vim.HostSystem or vim.vm
+        :return: ContainerView
+        """
+
+        container = self.api.content.viewManager.CreateContainerView(
+            root, vim_type, True)
+        view = container.view
+        container.Destroy()
+        return view
+
+    @staticmethod
+    def _create_filter_spec(pc, esxi_hosts, prop):
+        """
+        Build a filter spec (some kind of sql_query) for the property
+        collector.
+        See community samples
+        https://github.com/vmware/pyvmomi-community-samples/blob
+        /bc14f63065aa360ceca0cca477a8b271d582a090/samples/filter_vms.py
+
+        :param pc: The property_collector
+        :param esxi_hosts: a ContainerView of vim.HostSystem (Get with get_obj
+        :param prop: a single string or list of strings. Properties to
+        collect. Eg name, overallStatus...
+        :return: the created filter spec
+        """
+
+        if not isinstance(prop, list):
+            prop = [prop]
+
+        objSpecs = []
+        for host in esxi_hosts:
+            objSpec = vmodl.query.PropertyCollector.ObjectSpec(obj=host)
+            objSpecs.append(objSpec)
+        filterSpec = vmodl.query.PropertyCollector.FilterSpec()
+        filterSpec.objectSet = objSpecs
+        propSet = vmodl.query.PropertyCollector.PropertySpec(all=False)
+        propSet.type = vim.HostSystem
+        propSet.pathSet = prop
+        filterSpec.propSet = [propSet]
+        return filterSpec
+
+    def get_esxi_overall_stats(self):
+        """
+        Get the overall status of all esxi-hosts from vcenter.
+        Status can be green, yellow or red.
+
+        :return: a list of interfaces.Host
+        """
+
+        try:
+            self.login()
+            esxi = self._get_obj(self.api.content.rootFolder, [vim.HostSystem])
+            pc = self.api.content.propertyCollector
+            filter_spec = self._create_filter_spec(
+                pc, esxi, ['overallStatus', 'name'])
+            options = vmodl.query.PropertyCollector.RetrieveOptions()
+            result = pc.RetrievePropertiesEx([filter_spec], options)
+
+            res = []
+            vcenter = Vcenter(name=self.host, address=self.host)
+
+            state_map = {
+                'green': 2, 'yellow': 1, 'red': 0,
+            }
+
+            for item in result.objects:
+                name = [v.val for v in item.propSet if v.name == 'name'][0]
+                status = [v.val for v in item.propSet if v.name ==
+                          'overallStatus'][0]
+                host = Host(name=name, address=name, overall_status=state_map[status],
+                            vcenter=vcenter)
+                res.append(host)
+
+            return res

```

Comment:
Trying to catch everything here, we don't see exactly where the problem resides. This is bad to test and even worse to maintain.  Try to intercept only the disconnection. After that there should be no more problems.

Reply:
removed try_excepts, only remains at connection related code.